runtime/runc_jailer.go

```diff
@@ -0,0 +1,405 @@
+// Copyright 2018-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"syscall"
+
+	"github.com/firecracker-microvm/firecracker-go-sdk"
+	models "github.com/firecracker-microvm/firecracker-go-sdk/client/models"
+	"github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	"github.com/firecracker-microvm/firecracker-containerd/internal"
+	"github.com/firecracker-microvm/firecracker-containerd/internal/vm"
+)
+
+// runcJailer uses runc to set up a jailed environment for the Firecracker VM.
+type runcJailer struct {
+	ctx    context.Context
+	logger *logrus.Entry
+	// ociBundlePath is the path that will be used to create an OCI bundle,
+	// https://github.com/opencontainers/runtime-spec/blob/master/bundle.md
+	ociBundlePath string
+	// runcBinaryPath is the path used to execute the runc binary from.
+	runcBinaryPath string
+	uid            uint32
+	gid            uint32
+}
+
+func newRuncJailer(ctx context.Context, logger *logrus.Entry, ociBundlePath, runcBinPath string, uid, gid uint32) (*runcJailer, error) {
+	l := logger.WithField("ociBundlePath", ociBundlePath).
+		WithField("runcBinaryPath", runcBinPath)
+
+	j := &runcJailer{
+		ctx:            ctx,
+		logger:         l,
+		ociBundlePath:  ociBundlePath,
+		runcBinaryPath: runcBinPath,
+		uid:            uid,
+		gid:            gid,
+	}
+
+	rootPath := j.RootPath()
+
+	const mode = os.FileMode(0700)
+	// Create the proper paths needed for the runc jailer
+	j.logger.WithField("rootPath", rootPath).Debug("Creating root drive path")
+	if err := mkdirAndChown(rootPath, mode, j.uid, j.gid); err != nil {
+		return nil, errors.Wrapf(err, "%s failed to mkdirAndChown", rootPath)
+	}
+
+	return j, nil
+}
+
+// JailPath returns the base directory from where the jail binary will be ran
+// from
+func (j runcJailer) OCIBundlePath() string {
+	return j.ociBundlePath
+}
+
+// RootPath returns the root fs of the jailed system.
+func (j runcJailer) RootPath() string {
+	return filepath.Join(j.OCIBundlePath(), rootfsFolder)
+}
+
+// JailPath will return the OCI bundle rootfs path
+func (j runcJailer) JailPath() vm.Dir {
+	return vm.Dir(j.RootPath())
+}
+
+// BuildJailedMachine will return the needed options for a jailed Firecracker
+// instance. In addition, some configuration values will be overwritten to the
```

Comment:
"some configuration values" -> "Both cfg and machineConfig" to clarify what it does. BTW, the function is modifying the former. Right?

Reply:
Yea